kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java

```diff
@@ -0,0 +1,172 @@
+
+        createContainer(CONTAINER_1_ID, RELEASE_ID);
+    }
+
+    @Override
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses) throws Exception {
+        extraClasses.put(CLASS_CLOUD_BALANCE, Class.forName(CLASS_CLOUD_BALANCE, true, kieContainer.getClassLoader()));
+        extraClasses.put(CLASS_CLOUD_COMPUTER, Class.forName(CLASS_CLOUD_COMPUTER, true, kieContainer.getClassLoader()));
+        extraClasses.put(CLASS_CLOUD_PROCESS, Class.forName(CLASS_CLOUD_PROCESS, true, kieContainer.getClassLoader()));
+    }
+
+    @Test
+    public void testSolverWithAsyncResponseHandler() throws Exception {
+        Marshaller marshaller = MarshallerFactory.getMarshaller(new HashSet<Class<?>>(extraClasses.values()),
+                configuration.getMarshallingFormat(), client.getClassLoader());
+        ResponseCallback responseCallback = new BlockingResponseCallback(marshaller);
+        ResponseHandler asyncResponseHandler = new AsyncResponseHandler(responseCallback);
+        solverClient.setResponseHandler(asyncResponseHandler);
+
+        assertNoResponse(solverClient.createSolver(CONTAINER_1_ID, SOLVER_1_ID, SOLVER_1_CONFIG));
+        SolverInstance solver = responseCallback.get(SolverInstance.class);
+        assertThat(solver).isNotNull();
+        assertThat(solver.getContainerId()).isEqualTo(CONTAINER_1_ID);
+        assertThat(solver.getSolverId()).isEqualTo(SOLVER_1_ID);
+
+        assertNoResponse(solverClient.getSolvers(CONTAINER_1_ID));
+        SolverInstanceList solverList = responseCallback.get(SolverInstanceList.class);
+        assertThat(solverList).isNotNull();
+        assertThat(solverList.getContainers()).isNotNull().isNotEmpty().hasSize(1);
+        solver = solverList.getContainers().get(0);
+        assertThat(solver.getSolverId()).isEqualTo(SOLVER_1_ID);
+        assertThat(solver.getStatus()).isEqualTo(SolverInstance.SolverStatus.NOT_SOLVING);
+
+        solver = new SolverInstance();
+        solver.setStatus(SolverInstance.SolverStatus.SOLVING);
+        solver.setPlanningProblem(loadPlanningProblem(5, 15));
+        assertNoResponse(solverClient.updateSolverState(CONTAINER_1_ID, SOLVER_1_ID, solver));
+        solver = responseCallback.get(SolverInstance.class);
+        assertThat(solver.getSolverId()).isEqualTo(SOLVER_1_ID);
+        assertThat(solver.getStatus()).isEqualTo(SolverInstance.SolverStatus.SOLVING);
+
+        assertNoResponse(solverClient.getSolverState(CONTAINER_1_ID, SOLVER_1_ID));
+        solver = responseCallback.get(SolverInstance.class);
+        assertThat(solver.getSolverId()).isEqualTo(SOLVER_1_ID);
+        assertThat(solver.getStatus()).isEqualTo(SolverInstance.SolverStatus.SOLVING);
+
+        solverClient.setResponseHandler(new RequestReplyResponseHandler());
+        KieServerSynchronization.waitForSolverStatus(solverClient, CONTAINER_1_ID, SOLVER_1_ID, SolverInstance.SolverStatus.NOT_SOLVING);
```

Comment:
I think it should wait until the solving is finished in order to dispose the solver.  Do you think this assumption is wrong and we can skip it in order to save some time?